core/src/main/java/org/elasticsearch/plugins/SearchPlugin.java

```diff
@@ -99,6 +109,38 @@ public ScoreFunctionSpec(String name, Reader<T> reader, ScoreFunctionParser<T> p
     }
 
     /**
+     * Specification of custom {@link Query}.
+     */
+    public class QuerySpec<T extends QueryBuilder> extends SearchPluginSpec<T, QueryParser<T>> {
```

Comment:
can't we simply return `SearchPluginSpec` instead of having yet another  class for this? Maybe call it `SearchExtensionSpec` instead